---
title: "Untitled"
output: html_document
date: "2023-08-07"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
install.packages("tidyverse")
library(tidyverse)
```

# 1)	Install packages in R
```{R}
install.packages("MendelianRandomization")
library(MendelianRandomization)
```

# 2)	SNPs to exposure
* Create a CSV file of strong genetic predictors of lead from Supplementary Table 1 of this paper https://www.ncbi.nlm.nih.gov/pubmed/25820613.
* Only use the rows with the rsid highlighted in yellow. Use Table 1 of the same paper to identify the effect allele. Use the following self-explanatory column headings (chrpos, other_allele, effect_allele, eaf, beta, se, gene, SNP). Make sure the letters in the effect allele column are in upper case. Upload this file to RStudio Cloud as lead.txt, i.e., click on Upload at the lower left. 


# 4)	Obtain estimates
Continue using program leadIHD-mez to
i.	Read your lead.txt file into R called “vtas” and format it
ii.	Read information about the associations of your genetic instruments with the outcome CAD from the file” cad-extract.csv”, we prepared from Cardiogram
iii.	Rename the columns in the outcome file to be the same for every file
iv.	Merge SNP to exposure and SNP to outcome files
v.	Align the SNPs on the same effect allele for exposure and outcome and have a look at the alignment.

```{R}
# remove everything from workspace
rm(list = ls(all = TRUE))
install.packages(c("MendelianRandomization",
                   "metafor"))
#tell R to use the packages you installed
library(MendelianRandomization)
library(metafor)
```

## 1. check it is the correct file with the correct column names
```{R}
# read snps to exposure into vtas
vtas <- read.table("lead.txt",
                   header = TRUE) #read the file you created and uploaded

vtas[1,]
nrow(vtas)
```

# 2. read snps to the outcome file 
```{R}
sl <- read.csv("cad-extract.csv")

sl[1,]
```

## 3. rename the columns in the outcome file to be the same for every file
```{R}
snpto <- data.frame(SNP = sl$markername,
                    beta.outcome = sl$beta,
                   se.outcome = sl$se_dgc,
                   effect_allele.outcome = sl$effect_allele,
                   other_allele.outcome =sl$noneffect_allele,
                   eaf.outcome = sl$effect_allele_freq,
                   pval.outcome = sl$p_dgc,
                   chrpos.outcome = paste(sl$chr,":",sl$bp_hg19,sep=""))

snpto[1,]
```

## 4. Merge SNP to exposure and SNP to outcome files
```{R}
dat <- merge(snpto,
             vtas,
             by = "SNP",
             all.x = TRUE) #merge SNP to exposure and SNP to outcome file
nrow(dat)
dat[1,]
```

## 5. Align the SNPs on the same effect allele for exposure and outcome
```{R}
# by changing sign of beta.exposure if effect alleles do not match
dat %>% 
  mutate(swapt = if_else(effect_allele.outcome == toupper(effect_allele),
                         1, -1)) %>% 
  mutate(beta = beta * swapt) -> dat2 # reverse beta for exposure t if effect alleles do not match


# have a look at the alignment, are there any palindromic SNPs?
# are they aligned correctly?
dat2 %>% 
  select(SNP, effect_allele, other_allele,
         effect_allele.outcome, other_allele.outcome,
         swapt, eaf.outcome)
```

## 6. get forest plot with fixed effects, same as Mendelianrandomization IVW with fixed effects 
```{R}
read_delim("SNP_Symbol.txt") -> SNP_gene_tbw

dat %>% 
  mutate(Wald = beta.outcome/beta) %>% # #Wald estimate
  mutate(Waldvar = se.outcome^2 / beta^2) %>% 
  inner_join(SNP_gene_tbw) %>% 
  mutate(lab = paste(SNP, gene, sep= " ")) %>% 
  select(SNP, Wald, lab, Waldvar) -> dat3

metafor::rma.uni(yi = dat3$Wald,
        vi = dat3$Waldvar,
        slab = dat3$lab,
        method = "FE") -> dmres
```


```{R}
png("Lab1/Forest.png",
    res = 300, w = 2000, h = 1500)
metafor::forest(dmres,
                 atransf = exp,
                 xlab = " ",
                 mlab = "Ischemic heart disease (OR)",
                 at = log(c(.5, 1, 2)),
                 xlim = c(-1.7, 1.3),
                 cex = .8)
dev.off()
```


# 6. Get estimates for ihd using Mendelian randomization package with fixed effects
```{R}
mr_ivw <- MendelianRandomization::mr_ivw
mr_median <- MendelianRandomization::mr_median

library(MendelianRandomization)

mr_input(dat$beta,
         dat$se,
         dat$beta.outcome,
         dat$se.outcome) -> MRInputObject

mr_ivw(MRInputObject,
       model = "fixed") -> results

data.frame(Estimate = results@Estimate,
           StdError = results@StdError) %>% 
  mutate(LCI = Estimate - 1.96 * StdError) %>% 
  mutate(UCI = Estimate + 1.96 * StdError) %>% 
  mutate_all(funs(round(., 2)))
```


## 8. Get MR estimates for ihd using random effects
```{R}
mr_ivw(MRInputObject) -> results

# exponentiate to get OR and 95% CIs matching the forest plot
data.frame(Estimate = results@Estimate,
           StdError = results@StdError) %>% 
  mutate(LCI = Estimate - 1.96 * StdError) %>% 
  mutate(UCI = Estimate + 1.96 * StdError) %>% 
  mutate_all(funs(round(., 2)))
```


## 9. save file for future use
```{R}
write_csv(dat,
          "leadihd21.csv")
```

# use MR base code to do the analysis
```{R}
devtools::install_github('MRCIEU/TwoSampleMR')
library(TwoSampleMR)
```

```{R}
ao <- available_outcomes()
```

```{R}
read_exposure_data(filename = 'lead.txt',
                    sep = ' ',
                    snp_col = 'SNP',
                    beta_col = 'beta',
                    se_col = 'se',
                    effect_allele_col = 'effect_allele',
                    phenotype_col = 'Phenotype',
                    units_col = 'units',
                    other_allele_col = 'other_allele',
                    eaf_col = 'eaf',
                    samplesize_col = 'samplesize',
                    ncase_col = 'ncase',
                    ncontrol_col = 'ncontrol',
                    gene_col = 'gene',
                    pval_col = 'pval') -> exposure_dat

extract_outcome_data(exposure_dat$SNP,
                     c('ieu-a-7'),
                     proxies = 1,
                     rsq = 0.8,
                     align_alleles = 1,
                     palindromes = 1,
                     maf_threshold = 0.3) -> outcome_dat

harmonise_data(exposure_dat,
               outcome_dat,
               action = 2) -> dat

mr(dat) -> mr_results
```

